exec.go

```diff
@@ -122,122 +206,388 @@ func (p *Cmd) Start() (err error) {
+// output and standard error.
 //
-// The returned error is nil if the command runs and exits with a zero exit status.
+// Taken from exec/exec.go.
+func (c *Cmd) CombinedOutput() ([]byte, error) {
+	if c.Stdout != nil {
+		return nil, errors.New("wsl: Stdout already set")
+	}
+	if c.Stderr != nil {
+		return nil, errors.New("wsl: Stderr already set")
+	}
+	var b bytes.Buffer
+	c.Stdout = &b
+	c.Stderr = &b
+	err := c.Run()
+	return b.Bytes(), err
+}
+
+// StdoutPipe returns a pipe that will be connected to the command's
+// standard output when the command starts.
 //
-// If the command fails to run or doesn't complete successfully, the error is of type ExitError.
-func (p *Cmd) Wait() (err error) {
+// Wait will close the pipe after seeing the command exit, so most callers
+// need not close the pipe themselves. It is thus incorrect to call Wait
+// before all reads from the pipe have completed.
+// For the same reason, it is incorrect to call Run when using StdoutPipe.
+//
+// Based on exec/exec.go.
+func (c *Cmd) StdoutPipe() (io.ReadCloser, error) {
+	if c.Stdout != nil {
+		return nil, errors.New("wsl: Stdout already set")
+	}
+	if c.handle != 0 {
+		return nil, errors.New("wsl: StdoutPipe after process started")
+	}
+	pr, pw, err := os.Pipe()
+	if err != nil {
+		return nil, err
+	}
+	c.Stdout = pw
+	c.closeAfterStart = append(c.closeAfterStart, pw)
+	c.closeAfterWait = append(c.closeAfterWait, pr)
+	return pr, nil
+}
+
+// StderrPipe returns a pipe that will be connected to the command's
+// standard error when the command starts.
+//
+// Wait will close the pipe after seeing the command exit, so most callers
+// need not close the pipe themselves. It is thus incorrect to call Wait
+// before all reads from the pipe have completed.
+// For the same reason, it is incorrect to use Run when using StderrPipe.
+//
+// Based on exec/exec.go.
+func (c *Cmd) StderrPipe() (io.ReadCloser, error) {
+	if c.Stderr != nil {
+		return nil, errors.New("wsl: Stderr already set")
+	}
+	if c.handle != 0 {
+		return nil, errors.New("wsl: StderrPipe after process started")
+	}
+	pr, pw, err := os.Pipe()
+	if err != nil {
+		return nil, err
+	}
+	c.Stderr = pw
+	c.closeAfterStart = append(c.closeAfterStart, pw)
+	c.closeAfterWait = append(c.closeAfterWait, pr)
+	return pr, nil
+}
+
+func (c *Cmd) stdin() error {
+	// TODO
+	return nil
+}
+
+func (c *Cmd) stdout() error {
+	w, e := c.writerDescriptor(c.Stdout)
+	if e == nil {
+		c.stdoutW = w
+	}
+	return e
+}
+
+func (c *Cmd) stderr() error {
+	// Case where Stdout and Stderr are the same
+	if c.Stderr != nil && interfaceEqual(c.Stdout, c.Stderr) {
+		c.stderrW = c.stdoutW
+		return nil
+	}
+	// Different stdout and stderr
+	w, e := c.writerDescriptor(c.Stderr)
+	if e == nil {
+		c.stderrW = w
+	}
+	return e
+}
+
+// interfaceEqual protects against panics from doing equality tests on
+// two interfaces with non-comparable underlying types.
+func interfaceEqual(a, b any) bool {
```

Comment:
>  will silently write to stderr only and not stdout?

Not really. I think this is just an optimization.

This is my understanding of Cmd Stdout/Stderr redirection:

- If Cmd.Stdout and Cmd.Stderr are not the same, we create a pipe for each in (*Cmd).stdout and (*Cmd).stderr:
```
stdout ──> Cmd.Stdout ──> Your stdout reader
stderr ──> Cmd.Stderr ──> Your stderr reader
```
- If Cmd.Stderr and Cmd.Stdout are comparable and equal, we reuse the the pipe created in `(*Cmd).stdout`:
```
stdout ─┬─> Cmd.Stdout ──> Your reader
stderr ─┘
```
- If Cmd.Stdout and Cmd.Stderr are the same, but of a non-comparable type, the following diagram describes the piping:
```
stdout ──> [   Cmd.Stdout  ]
           [ (same object) ] ──> Your reader
stderr ──> [   Cmd.Stderr  ]
```
We see how this is equivalent to the previous, just less efficient because you need two pipes and two goroutines to copy them instead of a single one.

Reply:
Ah, this is used in the pipe context to compare them, making sense then! Thanks for the detailed explanation.